File: Services/pkg/domain/permission.go
package domain

import (
	"context"

	"github.com/ChangerzaryX1602/SkillSync/pkg/models"

	helpers "github.com/zercle/gofiber-helpers"
)

type PermissionRepository interface {
	Migrate() error
	CreatePermission(ctx context.Context, permission models.Permission) *helpers.ResponseError
	GetPermission(ctx context.Context, id uint) (*models.Permission, *helpers.ResponseError)
	GetPermissions(ctx context.Context, pagination models.Pagination, search models.Search) ([]models.Permission, *models.Pagination, *models.Search, *helpers.ResponseError)
	UpdatePermission(ctx context.Context, id uint, permission models.Permission) *helpers.ResponseError
	DeletePermission(ctx context.Context, id uint) *helpers.ResponseError
}

type PermissionService interface {
	CreatePermission(ctx context.Context, permission models.Permission) []helpers.ResponseError
	GetPermission(ctx context.Context, id uint) (*models.Permission, []helpers.ResponseError)
	GetPermissions(ctx context.Context, pagination models.Pagination, search models.Search) ([]models.Permission, *models.Pagination, *models.Search, []helpers.ResponseError)
	UpdatePermission(ctx context.Context, id uint, permission models.Permission) []helpers.ResponseError
	DeletePermission(ctx context.Context, id uint) []helpers.ResponseError
}

File: hono-clean-architecture/pkg/domain/permission.ts
import { Permission, Pagination, Search } from "../models";
import { ResponseError } from "../../internal/infrastructure/custom_error";

export interface PermissionRepository {
  migrate(): Promise<void>;
  createPermission(permission: {
    group: string;
    name: string;
  }): Promise<{ permission: Permission | null; error: ResponseError | null }>;
  getPermission(
    id: number
  ): Promise<{ permission: Permission | null; error: ResponseError | null }>;
  getPermissions(
    pagination: Pagination,
    search: Search
  ): Promise<{
    permissions: Permission[];
    pagination: Pagination;
    search: Search;
    error: ResponseError | null;
  }>;
  updatePermission(id: number, permission: Partial<Permission>): Promise<ResponseError | null>;
  deletePermission(id: number): Promise<ResponseError | null>;
  getPermissionByGroupAndName(
    group: string,
    name: string
  ): Promise<{ permission: Permission | null; error: ResponseError | null }>;
}

export interface PermissionService {
  createPermission(permission: { group: string; name: string }): Promise<ResponseError[]>;
  getPermission(id: number): Promise<{ permission: Permission | null; errors: ResponseError[] }>;
  getPermissions(
    pagination: Pagination,
    search: Search
  ): Promise<{
    permissions: Permission[];
    pagination: Pagination;
    search: Search;
    errors: ResponseError[];
  }>;
  updatePermission(id: number, permission: Partial<Permission>): Promise<ResponseError[]>;
  deletePermission(id: number): Promise<ResponseError[]>;
}
